File: March-2023/C++/MinimumScoreOfAPathBetweenTwoCities.cxx
//Minimum Score of a Path Between Two Cities
// we cannot find the minimum distance in the array directly because the graph is not necessarily connected, which means that some cities might not be reachable from others. So, if we loop through the array and find the minimum distance, we might miss some cities and therefore not be able to find the minimum score of a path between cities 1 and n.
// Instead, we need to traverse the graph using a search algorithm, such as depth-first search (DFS) or breadth-first search (BFS), to find all the possible paths between cities 1 and n and calculate the minimum score of each path. This ensures that we visit all the cities and find the minimum score of a path between cities 1 and n in the entire graph.
class Solution {
public:

    void dfs(unordered_map<int, vector<pair<int, int>>> &adj, int u, vector<bool>& visited, int &result) {
        
        visited[u] = true;
        
        for(auto &vec : adj[u]) {
            int v = vec.first;
            int c = vec.second;
            
            result = min(result, c);
            
            if(!visited[v]) {
                dfs(adj, v, visited, result);
            }
        }
    }
    
    int minScore(int n, vector<vector<int>>& roads) {
        unordered_map<int, vector<pair<int, int>>> adj;
        
        for(auto &vec : roads) {
            
            int u = vec[0];
            int v = vec[1];
            int c = vec[2];
            
            adj[u].push_back({v, c});
            adj[v].push_back({u, c});
        }
        
        vector<bool> visited(n, false);
        int result = INT_MAX;
        dfs(adj, 1, visited, result);
        
        return result;
    }
};

File: March-2023/Java/MinimumScoreOfAPathBetweenTwoCities.java
// Minimum Score of a Path Between Two Cities
// we cannot find the minimum distance in the array directly because the graph is not necessarily connected, which means that some cities might not be reachable from others. So, if we loop through the array and find the minimum distance, we might miss some cities and therefore not be able to find the minimum score of a path between cities 1 and n.

// Instead, we need to traverse the graph using a search algorithm, such as depth-first search (DFS) or breadth-first search (BFS), to find all the possible paths between cities 1 and n and calculate the minimum score of each path. This ensures that we visit all the cities and find the minimum score of a path between cities 1 and n in the entire graph.

class Solution {
    int result = Integer.MAX_VALUE;
    
    public void dfs(HashMap<Integer,List<int[]>> map, boolean[] visited, int u){
        visited[u]=true;
        for(int[] adj: map.get(u)){
            int v = adj[0];
            int w = adj[1];
            result = Math.min(result,w);
            if(!visited[v]){
           // result = Math.min(result,w);
            dfs(map,visited,v);
            }                        
        }               
    }
        
    public int minScore(int n, int[][] roads) {      
        HashMap<Integer,List<int[]>> map = new HashMap<>();
        boolean[] visited = new boolean[n+1];
        for(int i=0; i<roads.length; i++)
        {
            int u = roads[i][0];
            int v = roads[i][1];
            int w = roads[i][2];
            if(!map.containsKey(u)){
                map.put(u,new LinkedList<>());
            }
            if(!map.containsKey(v)){
                map.put(v,new LinkedList<>());
            }
            map.get(u).add(new int[]{v,w});
            map.get(v).add(new int[]{u,w});            
        }
        dfs(map,visited,1);
        return result;
    }
}
